### ISSUE TITLE
Make this gem compatible with rack-attack

### ISSUE THREAD

**Issue by collimarco** [modosc/cloudflare-rails#7]
This gem works properly for logging, etc. However currently it does not work with the `rack-attack` gem.

```
Rack::Attack.throttle('requests per ip', limit: 300, period: 5.minutes) do |request|
  # the request object is a Rack::Request
  # https://www.rubydoc.info/gems/rack/Rack/Request
  request.ip unless request.path.start_with? '/assets/'
end
```

Here the `request.ip` is the Cloudflare IP and not the client IP. 

Also using `request.remote_ip` raises `NoMethodError (undefined method remote_ip for #<Rack::Attack::Request:...>)`.

**Comment by modosc** [comment]
are you sure this is getting loaded before rack::attack? can you check rake middleware’s output?

**Comment by collimarco** [comment]
@modosc This  is the result of running `RAILS_ENV=production bundle exec rake middleware`:

```
use UTF8Cleaner::Middleware
use ActionDispatch::SSL
use Rack::Sendfile
use ActionDispatch::Executor
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use RequestStore::Middleware
use Rack::Timeout
use ActionDispatch::RemoteIp
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use ActionDispatch::DebugExceptions
use ActionDispatch::Callbacks
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ContentSecurityPolicy::Middleware
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Rack::TempfileReaper
use Warden::Manager
use Rack::Attack
run MyApplication::Application.routes
```

**Comment by modosc** [comment]
hrm, do you have an example app you can point me to?

**Comment by collimarco** [comment]
Thanks for your help! The only relevant things in the app are including `cloudflare-rails` and `rack-attack` gems in the `Gemfile` and then use my [configuration above](https://github.com/modosc/cloudflare-rails/issues/7#issue-408548129) for throttling. You will see that the `request.ip` is the Cloudflare IP.

**Comment by modosc** [comment]
can you try out the [rack-2.0](https://github.com/modosc/cloudflare-rails/tree/rack-2.0) branch? it looks like prepending my patches to `Rack::Request::Helper` stopped working at some point in the near past :/

**Comment by 6temes** [comment]
I tried the rack-2.0 branch, and it does not seem to help. `request.remote_ip` does not exist, and `request.ip` corresponds to the Cloudflare IP.

**Comment by modosc** [comment]
ok - i figured out the issue:
```
  # the request object is a Rack::Request
  # https://www.rubydoc.info/gems/rack/Rack/Request
```
this is not the case - it's actually a `Rack::Attack::Request` which subclasses `Rack::Request` and does so before this gem does all of its monkeypatching. 

i added a workaround for this and specs to handle it, this is released as [0.5.0](https://github.com/modosc/cloudflare-rails/releases/tag/v0.5.0). sorry for the long wait here!
